packages/server/src/sdk/app/tables/internal/index.ts

```diff
@@ -0,0 +1,187 @@
+import {
+  RenameColumn,
+  Table,
+  ViewStatisticsSchema,
+  ViewV2,
+  Row,
+  ContextUser,
+} from "@budibase/types"
+import {
+  hasTypeChanged,
+  TableSaveFunctions,
+} from "../../../../api/controllers/table/utils"
+import { FieldTypes } from "../../../../constants"
+import { EventType, updateLinks } from "../../../../db/linkedRows"
+import { cloneDeep } from "lodash/fp"
+import isEqual from "lodash/isEqual"
+import { runStaticFormulaChecks } from "../../../../api/controllers/table/bulkFormula"
+import { context } from "@budibase/backend-core"
+import { getTable } from "../getters"
+import { checkAutoColumns } from "./utils"
+import * as viewsSdk from "../../views"
+import sdk from "../../../index"
+import { getRowParams } from "../../../../db/utils"
+import { quotas } from "@budibase/pro"
+import env from "../../../../environment"
+import { cleanupAttachments } from "../../../../utilities/rowProcessor"
+
+export async function save(
+  table: Table,
+  opts?: {
+    user?: ContextUser
+    tableId?: string
+    rowsToImport?: Row[]
+    renaming?: RenameColumn
+  }
+) {
+  const db = context.getAppDB()
+
+  // if the table obj had an _id then it will have been retrieved
+  let oldTable: Table | undefined
+  if (opts?.tableId) {
+    oldTable = await getTable(opts.tableId)
+  }
+
+  // check all types are correct
+  if (hasTypeChanged(table, oldTable)) {
+    throw new Error("A column type has changed.")
```

Comment:
Would be nice in a follow-up PR to expand on this error. Looking at the implementation in `hasTypeChanged`, it might be worth throwing the error from without `hasTypeChanged` and make its name `checkColumnTypes` or something.

Reply:
Agreed - theres a lot of improved validation and error messaging we could do around the table API.